### DISCUSSION TITLE
Support for GCP Cloud Storage

### DISCUSSION THREAD

**Discussion by KaoruDev** [outline/outline#2440] [Ideas]
Hi I was wondering if there are any plans to support GCP Cloud Storage as a document storage backing.

**Comment by tommoor** [comment]
Hi, no plans to build this separately – GCP is compatible with the S3 API's:

https://cloud.google.com/storage/docs/migrating#migration-simple

**Reply by KaoruDev** [reply]
ah nice, TIL thank you!

**Reply by danieljrich** [reply]
I just tried implementing this, and came so close! I am able to successfully use Minio as a datastore. And when updating the environment variables to reference Google Cloud Storage, attaching an image _almost_ works. I say almost, because it does land in the bucket, but after it is uploaded, attachements.redirect tries to send me to `https://s3.us-east4.amazonaws.com`.
Even stranger, if I attach while minio is the datastore, and then change it to GCS, attachments.redirect now send me to the s3 url.
to add to the strangeness again, if I mistype or invalidate the gcs url (`https://wrong-bucket.storage.googleapis.com)`, attachments.redirect doesn't assume the s3 url.

I haven't dug into the code too much, but it's almost as if AWS_S3_FORCE_PATH_STYLE = true isn't being respected all the time.

**Reply by tommoor** [reply]
I kind of know why this is, there is some awkward code in Outline which tries to account for both path-style and non-path-style s3 urls that you must be hitting:

https://github.com/outline/outline/blob/83a61b87ed59c4c61dda20a7fe71f37e8aa7faa6/server/utils/s3.js#L20-L23

Is it possible to config so that the `AWS_S3_UPLOAD_BUCKET_URL` does not include the bucket name?

**Comment by rajesh-arunachalam** [comment]
It is possible to have a separate GCP Cloud Storage component and configure it through env.  Shall I contribute to building this feature?

**Reply by tommoor** [reply]
As mentioned previously we're not interested in maintaining multiple codepaths for this – GCP can already be used through their interoperability feature.
